doc/config.md
# Config File

When compiling a project, the software will look for `spec.toml` in the current directory.
A default config file will be created when the project is first ran, which should look something like this:

```toml
database = "spec.db"
document = "document.tex"
siteTitle = "Unnamed Website"

[compiler]
compileAll = false
redoTags = false
indirectReferences = true

[website]
font = "cmu-serif"
fontSize = 16
lineHeight = 1.3
textAlign = "left"
lineWidth = 45

primaryColour = "blue"
neutralColour = "grey"

searchLimit = 16
maxSearchPages = 48

recentChanges = 10
tableOfContentsDepth = 2

hoverPreview = true

displayComments = true
allowComments = true

advertiseSpec = true
```

## Shared Config Fields

- `database`: Location of the sqlite database file used by the project.
- `document`: Entry point of the LaTeX document.
- `siteTitle`: Title of the website, used for page titles and as a header for the main page.


## Compiler Config Fields

- `compileAll`: Whether to compile every tag whenever the compiler is ran. If set to `false`, the compiler will
  recognise existing labelled tags that have not been changed, and avoid rendering them again. Currently, setting
  `compileAll` to `true` should not significantly slow down the  However, more features in the future may change
  this.
- `redoTags`: If set to `true`, will **delete all existing tags** and assign new tags to every
  part/chapter/section/theorem/etc. This will break all existing links going into the website. Please consider taking a
  backup of the database file before doing this.
- `indirectReferences`: If set to `true`, will compute for each page the list of all pages it indirectly refers
  to. May slow down the compiler significantly when there is a large number of tags.


## Website Configs

- `font`: The main font used on the website. Must be one of `roboto`
  ([Roboto](https://fontsource.org/fonts/roboto)), `open-sans` ([Open Sans](https://fontsource.org/fonts/open-sans)),
  `cmu-serif` ([CMU Serif](https://fontlibrary.org/en/font/cmu-serif), default serif font in LaTeX), and
  `cmu-sans-serif` ([CMU Sans Serif](https://fontlibrary.org/en/font/cmu-sans-serif)).
- `fontSize`: Size of text in content blocks, in `px`.
- `lineHeight`: Height of each line.
- `textAlign`: Alignment of context texts. Must be one of `left`, `center`, `right`, `justify`.
- `lineWidth`: Width of the main content block, in `rem` (which scales with font size).


- `primaryColour`: Used to colour links and buttons. Must be one of `red`, `orange`, `amber`, `yellow`, `lime`, `green`, `emerald`, `teal`, `cyan`, `sky`,
  `blue`, `indigo`, `violet`, `purple`, `fuchsia`, `pink`, and `rose`.
- `neutralColour`: Used to colour background, lines, and shade elements. Must be one of `slate`, `grey`, `zinc`,
  `stone`. These colours can hardly be distinguished in light mode, but are more different in dark mode.


- `searchLimit`: Maximum amount of search results to display per page.
- `maxSearchPages`: Maximum number of pages for searches.


- `recentChanges`: The number of recent changes to show on the sidebar of the main page. If set to `0`, the recent changes section will never show up.


- `tableOfContentsDepth`: How many additional layers to display in the table of contents. For example, setting it to `1` will display all chapters on the main page, setting it to `2` will display all sections on the main page, and so on.
- `tableOfContentsUnfoldDepth`: Layers behind this number will start off folded in the table of contents. 
- `tableOfContentsShortDepth`: Layers behind this number will omit their name (section/subsection/etc) in the table of
contents. 


- `hoverPreview`: When hovering links, preview the target.
- `copyLabelButton`: Add a button next to theorems to copy their label.


- `displayComments`: Display comments under each unit.
- `allowComments`: Allow visitors to put down comments.


- `advertiseSpec`: Attribute spec in the sidebar.

The colour scheme is taken from [Tailwind](https://tailwindcss.com/docs/colors), thanks
to [HTML Colour Codes](https://htmlcolorcodes.com/color-chart/tailwind-color-chart/). The `grey` here refers to the
`neutral` in Tailwind.

To use a favicon, place the icon under `public/favicon.ico`. 



changelog.md
# Unreleased

### Bug Fixes

- References should now work in conservative mode. 
- Itemize environments now display their markers within enumerate environments. 
- Enumerate items now have proper paragraph spacing.


# v0.3.0, 2026-04-23

This update introduces a handful of breaking changes and new configurations. Please review the 
[config list](./doc/config.md) to understand their behaviour. If you experience a crash on the first compile attempt,
consider deleting the old database file. 

### Features

- Added the comments system. 
- Added customisation for the display and collapse of tables of contents via the configurations 
`tableOfContentsUnfoldDepth` and `tableOfContentsShortDepth`. 
- Reorganised the usage documentation to not remain on one page. 

### Bug Fixes

- Watch mode now refreshes pages after compile instead of before.
- `itemize` environments now have the same styling as `enumerate`s.
- The `documentclass` macro is now hidden from the main page.
- Fixed a bug where enumerates require an additional line break after.
- Fixed a bug where Firefox would not load XyPic properly due to not trusting a non-npm jsdelivr link. 
- Fixed a bug where using hyphens in the search bar crashes the website.
- Fixed a bug where conservative mode does not recognise custom environments.
- Fixed a bug where custom environments do not accept its optional argument. 
- The "recent changes" feature will now tracks when a unit is changed rather than when it is compiled.


# v0.2.3, 2026-04-23

- Added limited support for tables. 
- Added support for custom labels and start indices for enumerate environments. 
- Added support for arbitrary macro usage in labels. 


# v0.2.2, 2026-04-12

- Configs are no longer loaded multiple times in a row.
- Fixed a bug where theorem titles do not show up when it consists exclusively of macros. 
- Adjusted the npm index task. 

# v0.2.1, 2026-04-02

- Fixed MathJax paragraph rendering problem. 
- Added support for hover preview for units. 
- Fixed a bug where large documents crash on upsert. 
- Added a button next to theorems to copy their label.
- Conservative mode no longer overwrites the main page.
- Conservative mode no longer collects the preamble. 
- Added footnote support. 
- Enabled support for tikz pictures. 
- Allowed referencing inside math environments. 


# v0.2.0, 2026-03-25

- Added support for equation referencing. 
- Added automatic reloading on changes in watch mode. 
- Adjusted spacing of `\item` content. 
- Added font size, line height, and alignment config options.
- Units now get an automatically generated label based on its numbering if none is given. 
- Added the conservative mode to the `watch` command.
- Added config option to adjust the depth of the table of contents. 
- Added basic image inclusion.

# v0.1.6, 2026-03-19

- Fixed a bug where the QED square is placed improperly in some situations.
- Included additional metadata when loading.


# v0.1.5, 2026-03-13

- Slightly nudged some element spacing.
- Fixed a bug where the primary colour config does not apply. 
- Fixed a bug where unknown refs crash the compiler. 
- Updated light theme link colouring.


# v0.1.4, 2026-03-12

- Added a limit on searches. 
- Added support for TikZ figures.
- Fixed an inconsistency in search bar styling.
- Added self-promotion. 
- Added version self-reporting.
- Slightly adjusted block rendering. 


# v0.1.3, 2026-03-11

- Removed warnings for missing equation labels.
- Added support for tex commands in link texts.


# v0.1.2, 2026-02-28

- Fixed a bug where theorem counters are not accounted for properly.
- Fixed a bug where `document` environments are not expanded correctly.
- Adjusted page width and text spacing.
- XyJax diagrams now render somewhat properly in dark mode. 
- Block environment titles and their content now start in the same paragraph.
- Added a bit of documentation on commutative diagrams. 
- Added the `watch` command.

# v0.1.1, 2026-02-27

- Fixed a bug where the website is not initialised properly. 
- Adjusted the rendering of `\enumerate` labels. 
- Added documentation on custom packages and file imports.

# v0.1.0, 2026-02-27

Initial release. 

src/config.ts
import {z} from "zod";


export const SpecConfigSchema = z.object({
    database: z.string().default('spec.db'),
    document: z.string().default('document.tex'),
    siteTitle: z.string().default('Unnamed Website'),

    compiler: z.object({
        compileAll: z.boolean().default(false),
        redoTags: z.boolean().default(false),
        indirectReferences: z.boolean().default(true),
    }).default({}),

    website: z.object({
        font: z.enum(['roboto', 'open-sans', 'cmu-serif', 'cmu-sans-serif']).default('cmu-serif'),
        fontSize: z.number().int().min(1).default(16),
        lineHeight: z.number().default(1.3),
        textAlign: z.enum(['left', 'center', 'right', 'justify']).default('left'),
        lineWidth: z.number().int().min(15).max(120).default(45),

        primaryColour: z.enum(['red', 'orange', 'amber', 'yellow', 'lime', 'green', 'emerald', 'teal', 'cyan', 'sky',
            'blue', 'indigo', 'violet', 'purple', 'fuchsia', 'pink', 'rose']).default('blue'),
        neutralColour: z.enum(['slate', 'grey', 'zinc', 'stone']).default('grey'),

        searchLimit: z.number().int().min(1).default(16),
        maxSearchPages: z.number().int().min(1).default(48),

        recentChanges: z.number().int().min(0).max(32).default(10),
        tableOfContentsDepth: z.number().int().min(0).max(4).default(2),
        tableOfContentsUnfoldDepth: z.number().int().min(0).max(4).default(2),
        tableOfContentsShortDepth: z.number().int().min(0).max(4).default(1),

        hoverPreview: z.boolean().default(true),
        copyLabelButton: z.boolean().default(false),

        displayComments: z.boolean().default(false),
        allowComments: z.boolean().default(false),

        advertiseSpec: z.boolean().default(true),
    }).default({}),
});

export type SpecConfig = z.infer<typeof SpecConfigSchema>;

export const defaultConfig: SpecConfig = SpecConfigSchema.parse({});

export const defaultConfigPath = './spec.toml'

export let config: SpecConfig;

export function setConfig(newConfig: SpecConfig) {
    config = newConfig;
}
